src/routers/controllers/removeAuthorisedPersonRequestController.ts

```diff
@@ -0,0 +1,51 @@
+import { Request, Response } from "express";
+import * as constants from "../../constants";
+import { getExtraData, getLoggedInUserEmail, setExtraData } from "../../lib/utils/sessionUtils";
+import { Removal } from "../../types/removal";
+import { getCompanyAssociations, removeUserFromCompanyAssociations } from "../../services/associationsService";
+import { Associations } from "private-api-sdk-node/dist/services/associations/types";
+import logger from "../../lib/Logger";
+import { Session } from "@companieshouse/node-session-handler";
+import { CompanyNameAndNumber } from "../../types/util-types";
+import { validateRemoveAssociation } from "../../lib/validation/validateRemoveAssociation";
+
+/**
+ * This controller handles the removeUserFromCompanyAssociations (updateAssociationStatus) api call which
+ * updates the status for the association.
+ * If the user is removing themselves, it redirects them to the removing themselves success page, if removing
+ * another user it redirects to the manage authorised person page with a success message.
+ * If the request is invalid, the function throws an error
+ */
+export const removeAuthorisedPersonRequestController = async (req: Request, res: Response): Promise<void> => {
+
+    const companyNumber: string = req.params[constants.COMPANY_NUMBER];
+    const removal: Removal = getExtraData(req.session, constants.REMOVE_PERSON);
+
+    const companyAssociations: Associations = await getCompanyAssociations(req, companyNumber);
+
+    const associationToBeRemoved = companyAssociations.items.find(assoc => assoc.userEmail === removal.userEmail);
+
+    if (!associationToBeRemoved || !validateRemoveAssociation(associationToBeRemoved, removal, companyNumber)) {
+        throw new Error("validation for removal of assocation failed");
+    }
+
+    logger.info(`Request to remove association ${associationToBeRemoved.id}, ${associationToBeRemoved.companyNumber}`);
+
+    await removeUserFromCompanyAssociations(req, associationToBeRemoved.id);
```

Comment:
Should we check the result of this function and redirect to a relevant error message page given this fails?  At the moment I see a scenario where this promise is rejected, but the code after this block stills works because we don't check the result of it

Reply:
I don’t think the code block after this will run if the promise rejects. If a promise is rejected it’s similar to an error, 
the control jumps to the closest rejection handler.
In other words it will find the next part in the chain that can handle or catch the error, the rest is skipped. So normally in express async calls are wrapped in a try...catch block, the error is passed to next so it can be processed.
Like this…. 
    try {
	await removeUserFromCompanyAssociations(req, associationToBeRemoved.id);
    } catch (err) {
      next(err)
    }
We added a module called express-async-errors so we don’t require this try catch.
We have a central error handler which catches errors. Based on the error it can return a specific error page to the user. I think we do have upcoming tickets for this, we’re just returning a generic error template when this happens at the minute.

Reply:
Ah!  I'd thought something like this might've happened, so the rejection bubbles up and gets caught!  Nice one, all good then - thanks for detailing how that works too btw.